lib/src/ioctl/g_ext_ctrls.rs

```diff
@@ -0,0 +1,418 @@
+    H264Sps,
+    H264Pps,
+    H264ScalingMatrix,
+    H264PredWeights,
+    H264SliceParams,
+    H264DecodeParams,
+    FwhtParams(FwhtParams),
+}
+
+impl TryFrom<CompoundControl> for ExtControl {
+    type Error = DataConversionError;
+
+    fn try_from(ctrl: CompoundControl) -> Result<Self, Self::Error> {
+        match ctrl {
+            CompoundControl::H264Sps(_) => todo!(),
+            CompoundControl::H264Pps(_) => todo!(),
+            CompoundControl::H264ScalingMatrix(_) => todo!(),
+            CompoundControl::H264PredWeights(_) => todo!(),
+            CompoundControl::H264SliceParams(_) => todo!(),
+            CompoundControl::H264DecodeParams(_) => todo!(),
+            CompoundControl::FwhtParams(inner) => Ok(ExtControl::FwhtParams(FwhtParams {
+                backward_ref_ts: inner.backward_ref_ts,
+                version: inner.version,
+                width: inner.width,
+                height: inner.height,
+                flags: inner.flags,
+                components: (inner.flags & bindings::V4L2_FWHT_FL_COMPONENTS_NUM_MSK)
+                    >> bindings::V4L2_FWHT_FL_COMPONENTS_NUM_OFFSET,
+                encoding: (inner.flags & bindings::V4L2_FWHT_FL_PIXENC_MSK)
+                    >> bindings::V4L2_FWHT_FL_PIXENC_OFFSET,
+                colorspace: Colorspace::try_from(inner.colorspace)?,
+                xfer_func: XferFunc::try_from(inner.xfer_func)?,
+                ycbcr_enc: YCbCrEncoding::try_from(inner.ycbcr_enc)?,
+                quantization: Quantization::try_from(inner.quantization)?,
+            })),
+        }
+    }
+}
+
+#[doc(hidden)]
+mod ioctl {
+    use crate::bindings::v4l2_ext_controls;
+    nix::ioctl_readwrite!(vidioc_g_ext_ctrls, b'V', 71, v4l2_ext_controls);
+    nix::ioctl_readwrite!(vidioc_s_ext_ctrls, b'V', 72, v4l2_ext_controls);
+}
+
+#[derive(Debug, Error)]
+pub enum ExtControlError {
+    #[error("Unexpected ioctl error: {0}")]
+    IoctlError(nix::Error),
+    #[error("This control is not a compound control")]
+    NotACompoundControlError,
+    #[error("Failed data conversion: {0}")]
+    DataConversionError(DataConversionError),
+}
+
+impl From<DataConversionError> for ExtControlError {
+    fn from(err: DataConversionError) -> Self {
+        ExtControlError::DataConversionError(err)
+    }
+}
+
+/// Get a single extended control
+///
+/// While vidioc_g_ext_ctrls() accepts an array of controls,
+/// the kernel internally seems not to commit all the values atomically,
+/// and proceeds only until the first failure.
+///
+/// Given the above we provide an interface for querying a single control,
+/// as this greatly simplifies the code.
+pub fn g_ext_ctrl<F: AsRawFd, R: AsRawFd>(
+    fd: &F,
+    request_fd: Option<&R>,
+    ctrl_kind: ExtControlKind,
+) -> Result<ExtControl, ExtControlError> {
+    let mut control = bindings::v4l2_ext_control {
+        // SAFETY: ok to zero-fill this struct, the pointer it contains will be assigned to in this function
```

Comment:
Speaking from Rust point of view. But from V4L2's point of view we'll get a "bad address" if the said pointers are null. I'd say it is the very purpose of a safe abstraction to prevent this from happening.